### PR TITLE
[PWX-37026] Adding FBDAMultiTenancyBasicTest

### PR DESCRIPTION
**Pull request by vinayakshnd**
**What this PR does / why we need it**:
This PR automates below testcases:
1. https://portworx.testrail.net/index.php?/tests/view/87231737
2. https://portworx.testrail.net/index.php?/tests/view/87231739
3. https://portworx.testrail.net/index.php?/tests/view/87231756

Added `FBDAMultiTenancyBasicTest` testcase, which deploys app backed by FBDA volume and StorageClass for FBDA volume mentions `pure_nfs_endpoint` as parameter which can be different than one mentioned in pure.json. Test validates that volume uses endpoint mentioned in the storageClass. 

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

Successful torpedo runs: 
1. https://jenkins.pwx.dev.purestorage.com/job/Users/job/Vinayak/job/FBDA-MultiTenancy/22/
2. https://jenkins.pwx.dev.purestorage.com/job/Users/job/Vinayak/job/FBDA-MultiTenancy/23/

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -28,7 +28,7 @@ require (
 	github.com/kubernetes-incubator/external-storage v0.20.4-openstorage-rc7
 	github.com/libopenstorage/autopilot-api v1.3.0
 	github.com/libopenstorage/cloudops v0.0.0-20230220114907-3e63dce1b413
-	github.com/libopenstorage/openstorage v9.4.47+incompatible
+	//github.com/libopenstorage/openstorage v9.4.47+incompatible
```
Comment (nikolaypopov) [inline go.mod:31]: Remove it if its not needed, instead of commenting this out, though I still see it gets added as part of `require` on line 152

Comment (vinayakshnd) [reply go.mod]: Done. removed commented line and kept the required line 152. Yes, we need github.com/libopenstorage/openstorage v9.4.47+incompatible line in go.mod.

File: tests/basic/pure_test.go [hunk 2]
```diff
@@ -2810,3 +2813,71 @@ var _ = Describe("{VolAttachSameFAPxRestart}", func() {
...
+	var contexts []*scheduler.Context
+	var testName string
+	var customConfigAppName string
+
+	testName = "fbda-multitenancy"
+	JustBeforeEach(func() {
+		StartTorpedoTest("FBDAMultiTenancyBasicTest", "Validate FBDA vols get consumed over IP mentioned in `pure_nfs_endpoint` parameter of storageClass", nil, 0)
+		Step("setup credential necessary for cloudsnap", createCloudsnapCredential)
+
+		logrus.Infof("Using CustomAppConfig: %+v", Inst().CustomAppConfig)
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:2831]: please use log.infoF

File: tests/basic/pure_test.go [hunk 3]
```diff
@@ -2810,3 +2813,71 @@ var _ = Describe("{VolAttachSameFAPxRestart}", func() {
...
+			}
+		}
+	})
+
+	When("pure_nfs_endpoint parameter specified in storageClass", func() {
+		It("should create a FBDA volume over pure_nfs_endpoint mentioned in storageClass", func() {
+			ctx := findContext(contexts, customConfigAppName)
+
+			vols, err := Inst().S.GetVolumes(ctx)
+			Expect(err).NotTo(HaveOccurred())
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:2856]: please use dash.VerifyFatal for verifications

File: tests/basic/pure_test.go [hunk 4]
```diff
@@ -2810,3 +2813,71 @@ var _ = Describe("{VolAttachSameFAPxRestart}", func() {
...
+		JustAfterEach(func() {
+			opts := make(map[string]bool)
+			opts[scheduler.OptionsWaitForResourceLeakCleanup] = true
+
+			for _, ctx := range contexts {
+				TearDownContext(ctx, opts)
+			}
+			Step("delete credential used for cloudsnap", deleteCloudsnapCredential)
+
+			defer EndTorpedoTest()
```
Comment (kphalgun-px) [inline tests/basic/pure_test.go:2879]: Can we move this to immediately after JustAfterEach?

File: tests/basic/pure_test.go [hunk 5]
```diff
@@ -2810,3 +2813,71 @@ var _ = Describe("{VolAttachSameFAPxRestart}", func() {
...
+					found = true
+					customConfigAppName = appName
+					break
+				}
+			}
+			if !found {
+				logrus.Warnf("App %v not found in %d contexts, skipping test", appName, len(contexts))
+				Skip(fmt.Sprintf("app %v not found", appName))
+			}
+		}
```
Comment (kphalgun-px) [inline tests/basic/pure_test.go:2848]: Can we do this before scheduling the apps?

File: tests/basic/pure_test.go (hunk 2 above)
Comment (vinayakshnd) [reply tests/basic/pure_test.go]: done

File: tests/basic/pure_test.go (hunk 3 above)
Comment (vinayakshnd) [reply tests/basic/pure_test.go]: done

File: tests/basic/pure_test.go (hunk 4 above)
Comment (vinayakshnd) [reply tests/basic/pure_test.go]: done

File: tests/basic/pure_test.go (hunk 5 above)
Comment (vinayakshnd) [reply tests/basic/pure_test.go]: done

File: tests/basic/pure_test.go [hunk 6]
```diff
@@ -2810,3 +2813,71 @@ var _ = Describe("{VolAttachSameFAPxRestart}", func() {
...
+
+			vols, err := Inst().S.GetVolumes(ctx)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(vols)).To(Not(BeZero()))
+
+			expectedPureNfsEndpoint := Inst().CustomAppConfig[customConfigAppName].StorageClassPureNfsEndpoint
+
+			for _, vol := range vols {
+				apiVol, err := Inst().V.InspectVolume(vol.ID)
+				Expect(err).NotTo(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (vinayakshnd) [reply tests/basic/pure_test.go]: done

**Review by kphalgun-px — APPROVED** [review]
LGTM

File: tests/basic/pure_test.go [hunk 7]
```diff
@@ -2,8 +2,11 @@ package tests
 
 import (
 	"fmt"
+
 	"github.com/devans10/pugo/flasharray"
 	"github.com/portworx/sched-ops/k8s/storage"
+	"github.com/sirupsen/logrus"
```
[the remark this answers is not in the document]
Comment (vinayakshnd) [reply tests/basic/pure_test.go]: Done

File: tests/basic/pure_test.go [hunk 8]
```diff
@@ -2810,3 +2812,70 @@ var _ = Describe("{VolAttachSameFAPxRestart}", func() {
...
+			dash.VerifyNotNilFatal(err, "Failed to get list of volumes")
+			dash.VerifyFatal(len(vols) != 0, true, "Failed to get volumes")
+
+			expectedPureNfsEndpoint := Inst().CustomAppConfig[customConfigAppName].StorageClassPureNfsEndpoint
+
+			for _, vol := range vols {
+				apiVol, err := Inst().V.InspectVolume(vol.ID)
+				log.FailOnError(err, fmt.Sprintf("Failed to inspect volume [%s]", apiVol.GetId()))
+				// Validate the volume is created over the NFS endpoint mentioned in storageClass
+				Expect(apiVol.Spec.ProxySpec.PureFileSpec.NfsEndpoint).To(Equal(expectedPureNfsEndpoint), "FBDA volume is not using NFS endpoint mentioned in storageClass.")
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:2865]: use dash.verifyFatal

Comment (vinayakshnd) [reply tests/basic/pure_test.go]: Done

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Pull request merged**